### PR TITLE
TIP-521: Clean FlatItemBuffer

### PR DESCRIPTION
**Pull request by momoss**
FlatItemBuffer now extends JSONFileBuffer and is declared as a real factory

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | N
| Added Behats                      | N
| Changelog updated                 | Y
| Review and 2 GTM                  | Y
| Migration script                  | N
| Tech Doc                          | N

### CODE REVIEW

File: src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php [hunk 1]
```diff
@@ -61,9 +76,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $this->flatRowBuffer->write($products, $parameters->get('withHeader'));
-
-        $parameters = $this->stepExecution->getJobParameters();
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php]: why `$options['isWithHeader']` ?
$options has not been initilized and you don't use it in your code. 
you should inject `$parameters->get('withHeader')` directly in `write()` method

File: src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php [hunk 2]
```diff
@@ -66,8 +81,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $this->flatRowBuffer->write($variantGroups, $parameters->get('withHeader'));
-
+        $options['isWithHeader'] = $parameters->get('withHeader');
+        $this->flatRowBuffer->write($variantGroups, $options);
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php]: same

File: src/Pim/Component/Connector/Writer/File/Csv/Writer.php [hunk 3]
```diff
@@ -64,8 +79,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $isWithHeader = $parameters->get('withHeader');
-        $this->flatRowBuffer->write($items, $isWithHeader);
+        $options['isWithHeader'] = $parameters->get('withHeader');
+        $this->flatRowBuffer->write($items, $options);
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/Csv/Writer.php]: same

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php [hunk 4]
```diff
@@ -12,40 +12,31 @@
...
 
-    /**
-     * @param BufferFactory $bufferFactory
-     */
-    public function __construct(BufferFactory $bufferFactory)
+    public function __construct()
     {
-        $this->buffer = $bufferFactory->create();
-        $this->count  = 0;
+        parent::__construct();
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: useless

File: src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php]: you should init `options`

File: src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php [hunk 5]
```diff
@@ -66,8 +81,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $this->flatRowBuffer->write($variantGroups, $parameters->get('withHeader'));
-
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php]: same comment

File: src/Pim/Component/Connector/Writer/File/Csv/Writer.php [hunk 6]
```diff
@@ -64,8 +79,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $isWithHeader = $parameters->get('withHeader');
-        $this->flatRowBuffer->write($items, $isWithHeader);
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Csv/Writer.php]: same

File: src/Pim/Component/Connector/Writer/File/Xlsx/ProductWriter.php [hunk 7]
```diff
@@ -63,8 +77,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $withHeader = $parameters->get('withHeader');
-        $this->flatRowBuffer->write($products, $withHeader);
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Xlsx/ProductWriter.php]: same

File: src/Pim/Component/Connector/Writer/File/Xlsx/Writer.php [hunk 8]
```diff
@@ -54,8 +69,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $withHeader = $parameters->get('withHeader');
-        $this->flatRowBuffer->write($items, $withHeader);
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Xlsx/Writer.php]: same

File: src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php]: maybe call the option `withHeader`, just like it was in the JobParameters

**Comment by jjanvier** [comment]
GTM once fixed

**Comment by jjanvier** [comment]
(you should explain the PR in the description)

File: src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php [hunk 9]
```diff
@@ -16,7 +18,7 @@
  * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-class ProductWriter extends AbstractFileWriter implements ArchivableWriterInterface
+class ProductWriter extends AbstractFileWriter implements ItemWriterInterface, ArchivableWriterInterface
 {
     /** @var FlatItemBuffer */
     protected $flatRowBuffer;
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Csv/ProductWriter.php]: this should be initialized to null cf line 62

File: src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php [hunk 10]
```diff
@@ -30,28 +32,41 @@ class VariantGroupWriter extends AbstractFileWriter implements ArchivableWriterI
...
         $this->fileExporter = $fileExporter;
         $this->flusher = $flusher;
     }
 
     /**
      * {@inheritdoc}
      */
+    public function initialize()
+    {
+        if (null === $this->flatRowBuffer) {
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Csv/VariantGroupWriter.php]: idem than previous

File: src/Pim/Component/Connector/Writer/File/Csv/Writer.php [hunk 11]
```diff
@@ -31,23 +36,33 @@ class Writer extends AbstractFileWriter implements ArchivableWriterInterface
...
+        $this->bufferFactory = $bufferFactory;
         $this->flusher = $flusher;
     }
 
     /**
      * {@inheritdoc}
      */
+    public function initialize()
+    {
+        if (null === $this->flatRowBuffer) {
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Csv/Writer.php]: same

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php [hunk 12]
```diff
@@ -12,40 +12,31 @@
...
+class FlatItemBuffer extends JSONFileBuffer implements BufferInterface, \Countable
 {
-    /** @var BufferInterface */
-    protected $buffer;
-
     /** @var array */
     protected $headers = [];
 
     /** @var int */
     protected $count;
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: protected $count = 0; and you can remove the constructor.

File: src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php [hunk 13]
```diff
@@ -12,40 +12,31 @@
  * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-class FlatItemBuffer implements \Countable
+class FlatItemBuffer extends JSONFileBuffer implements BufferInterface, \Countable
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/FlatItemBuffer.php]: Not fan of that

File: src/Pim/Component/Connector/Writer/File/Xlsx/ProductWriter.php [hunk 14]
```diff
@@ -54,6 +58,16 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
+    public function initialize()
+    {
+        if (null === $this->flatRowBuffer) {
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Xlsx/ProductWriter.php]: same

File: src/Pim/Component/Connector/Writer/File/Xlsx/Writer.php [hunk 15]
```diff
@@ -26,26 +28,39 @@ class Writer extends AbstractFileWriter implements ArchivableWriterInterface
...
         $this->flusher       = $flusher;
         $this->writtenFiles  = [];
     }
 
     /**
      * {@inheritdoc}
      */
+    public function initialize()
+    {
+        if (null === $this->flatRowBuffer) {
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Xlsx/Writer.php]: same

File: src/Pim/Component/Connector/Writer/File/Xlsx/VariantGroupWriter.php [hunk 16]
```diff
@@ -68,8 +82,8 @@ public function write(array $items)
         }
 
         $parameters = $this->stepExecution->getJobParameters();
-        $withHeader = $parameters->get('withHeader');
-        $this->flatRowBuffer->write($variantGroups, $withHeader);
+        $options['isWithHeader'] = $parameters->get('withHeader');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Xlsx/VariantGroupWriter.php]: same

File: src/Pim/Component/Connector/Writer/File/Xlsx/VariantGroupWriter.php [hunk 17]
```diff
@@ -54,6 +58,16 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
+    public function initialize()
+    {
+        if (null === $this->flatRowBuffer) {
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Writer/File/Xlsx/VariantGroupWriter.php]: same

File: src/Akeneo/Component/Buffer/BufferInterface.php [hunk 18]
```diff
@@ -16,8 +16,9 @@
 {
     /**
      * @param mixed $item                   The item to write in the buffer
+     * @param array $options                The options required by the buffer
      *
      * @throws UnsupportedItemTypeException If the buffer implementation does not support item of this type
      */
-    public function write($item);
+    public function write($item, array $options = []);
```
Comment (skeleton) [inline src/Akeneo/Component/Buffer/BufferInterface.php]: bc break :)

**Comment by skeleton** [comment]
GTM once fixed

**Pull request merged**